docs/proposals/deletions_object_storage.md

```diff
@@ -0,0 +1,84 @@
+
+## Goals
+
+*   Unblock users and allow series deletion in the object storage using tombstones.
+*   Deal with compaction and downsampling of blocks with tombstones.
+*   Performing deletions at admin level.
+
+## Proposed Approach
+
+*   We propose to implement deletions using the tombstones approach.
+*   To start off we use the Prometheus tombstone file format.
+*   A user is expected to enter the following details for performing deletions:
+    *   **label matchers**
+    *   **start timestamp**
+    *   **end timestamp** (start and end timestamps of the series data the user expects to be deleted)
+*   The details are entered via a deletions API and they are processed by the compactor to create a tombstone file, if there’s a match for the details entered. Afterwards the tombstone file is uploaded to object storage making it accessible to other components.
+*   **Why Compactor**? : The compactor is one of the very few components (next to Sidecar, Receiver, Ruler) that has write capabilities to object storage. At the same time it is only ever running once per object storage, so that we have a single writer to the deletion tombstones.
+*   If the data with the requested details couldn’t be found in the storage an error message is returned back as a response to the request.
+*   Store Gateway masks the series on processing the tombstones from the object storage.
+*   **Perspectives to deal with Compaction of blocks having tombstones:**
+    *   **Block with tombstones** Have a threshold to perform deletions on the compacted blocks ([In Prometheus](https://github.com/prometheus/prometheus/blob/f0a439bfc5d1f49cec113ee9202993be4b002b1b/tsdb/compact.go#L213), the blocks with big enough time range, that have >5% tombstones, are considered for compaction.) We solve the tombstones, if the tombstones are greater than than the threshold and then perform compaction. If not we attach the tombstone file to the new block. If multiple blocks are being compacted, we merge the tombstone files of the blocks whose threshold is not met.
+    *   **Block with deletion-mark.json i.e., entire block marked for deletion:** Returns an error message as the entire block is going to be deleted.
+*   **Perspectives to deal with Downsampling of blocks:**
+    *   **Block with tombstones:** If the tombstones are less than the threshold we copy the tombstone file and attach it to the new downsampled block else we solve the tombstones and downsample the block. And the downsampled block with tombstones during its next compaction would again have the same cases as with the compaction of a block with tombstones.
+    *   **Blocks without tombstones:** Downsampling happens...
+
+##### Problems during implementations and possible ideas to solve :
+During the implementation phase of this proposal one of the first problems we came across was that we had to pull the index of all the blocks for creating or appending a tombstone.
```

Comment:
Yea, Compactor is not responsible for that. It does not load or need to load all the blocks in the system in order to craft tombstones. That's why we cannot use Prometheus format tombstone and have to figure out something else. If we would required to load indexes for ALL blocks in order to create deletion request - this will never scale. 

We need to delegate logic of figuring our which series ID to delete in leafs for querying and during compaction for rewrite.

The tricky part is to know which blocks have the data to be deleted. We thought about simple Series with `skipChunk` options